posts/2021-07-29-Rust-1.54.0.md

```diff
@@ -0,0 +1,78 @@
+---
+layout: post
+title: "Announcing Rust 1.54.0"
+author: The Rust Release Team
+release: true
+---
+
+# Announcing Rust 1.54.0
+
+The Rust team is happy to announce a new version of Rust, 1.54.0. Rust is a programming language empowering everyone
+to build reliable and efficient software.
+
+If you have a previous version of Rust installed via rustup, getting Rust
+1.54.0 is as easy as:
+
+```console
+rustup update stable
+```
+
+If you don't have it already, you can [get `rustup`][install]
+from the appropriate page on our website, and check out the
+[detailed release notes for 1.54.0][notes] on GitHub.
+
+[install]: https://www.rust-lang.org/install.html
+[notes]: XXX
+
+## What's in 1.54.0 stable
+
+### Attributes can call macros!()
+
+Rust 1.54 supports invoking function-like macros inside attributes. One notable use case for this is including documentation from other files into Rust doc comments. For example, if your project's README represents a good documentation comment, you can use `include_str!` to directly incorporate the contents. Previously, various workarounds allowed similar functionality, but from 1.54 this is much more ergonomic.
+
+
+```rust=
+#![doc = include_str!("README.md")]
+```
+
+Macros can be nested inside the attribute as well, for example to include content generated by a build script:
+
+```rust
+#[path = concat!(env!("OUT_DIR"), "/generated.rs")]
+mod generated;
+```
+
+Read [here](https://github.com/rust-lang/rust/pull/83366) for more details.
+
+### wasm32 intrinsics stabilized
+
+A number of intrinsics for the wasm32 platform have been stabilized, which gives access to the SIMD instructions in WebAssembly.
+
+Notably, unlike the previously stabilized `x86` and `x86_64` intrinsics, these do not have a safety requirement to only be called when the appropriate target feature is enabled. This is because WebAssembly was written from the start to validate code safely before executing it, so instructions are guaranteed to be decoded correctly (or not at all).
+
+This means that we can expose some of the intrinsics as entirely safe functions, for example [`v128_bitselect`](https://doc.rust-lang.org/beta/core/arch/wasm32/fn.v128_bitselect.html). However, there are still some intrinsics which are unsafe because they use raw pointers, such as [`v128_load`](https://doc.rust-lang.org/beta/core/arch/wasm32/fn.v128_load.html).
+
+### Incremental Compilation is XXX by default
+
+TODO - Wesley is going to draft text here.
+
```

Comment:
```suggestion
### Incremental Compilation is re-enabled by default

In Rust 1.52, additional validation was added when loading incremental compilation data from the on-disk cache.
This resulted in a number of pre-existing issues being uncovered as the validation changed these silent bugs into internal compiler errors (ICEs).
In response, the Compiler Team decided to disable incremental compilation in the 1.52.1 patch, allowing users to avoid encountering the ICEs. [^1]

Since then, we've conducted a [series of retrospectives][retros] and contributors have been hard at work resolving the reported issues with some fixes landing in 1.53 and the majority landing in this release. [^2]
There are currently still two known issues which can result in an ICE.
However, based on the feedback we received from users affected by the 1.52 release, we believe these issues to be rare.

Therefore, incremental compilation has been re-enabled in this release!

[^1]: The [1.52.1 release notes] contain a more detailed description of these events.
[^2]: The tracking issue for the issues is [#84970].

[#84970]: https://github.com/rust-lang/rust/issues/84970
[1.52.1 release notes]: https://blog.rust-lang.org/2021/05/10/Rust-1.52.1.html
[retros]: https://github.com/rust-lang/compiler-team/issues/435

```

cc @pnkfelix

Reply:
Nit: I'd avoid passive voice. E.g., instead of "This resulted in a number of pre-existing issues being uncovered ...", I'd say "This uncovered a number of pre-existing issues ..."

Reply:
In the compiler team meeting, we touched on the topic that the lack of telemetry data from the compiler impedes the feedback cycle: It makes it impossible for us to measure true impact, and also any feedback we get is relying on a human being to decide to report the issue. Did we want to make any mention of that here, as a way to explain how we are operating somewhat in the dark at this moment?

Reply:
I would lead with the good news.

```suggestion
### Incremental Compilation is re-enabled by default

Incremental compilation has been re-enabled by default in this release, after it being disabled by default in 1.52.

In Rust 1.52, additional validation was added when loading incremental compilation data from the on-disk cache.
This resulted in a number of pre-existing potential soundness issues being uncovered as the validation changed these silent bugs into internal compiler errors (ICEs).
In response, the Compiler Team decided to disable incremental compilation in the 1.52.1 patch, allowing users to avoid encountering the ICEs and the underlying unsoundness, at the expense of longer compile times. [^1]

Since then, we've conducted a [series of retrospectives][retros] and contributors have been hard at work resolving the reported issues, with some fixes landing in 1.53 and the majority landing in this release. [^2]

There are currently still two known issues which can result in an ICE.
Due to the lack of automated crash reporting, we can't be certain of the full extent of impact of the outstanding issues. However, based on the feedback we received from users affected by the 1.52 release, we believe these issues to be rare.

Therefore, incremental compilation has been re-enabled in this release!

[^1]: The [1.52.1 release notes] contain a more detailed description of these events.
[^2]: The tracking issue for the issues is [#84970].
[#84970]: https://github.com/rust-lang/rust/issues/84970
[1.52.1 release notes]: https://blog.rust-lang.org/2021/05/10/Rust-1.52.1.html
[retros]: https://github.com/rust-lang/compiler-team/issues/435
```